### PR TITLE
feat(crypto): CRP-2624 Create tECDSA/tSchnorr/vetKD mock canister

### PR DESCRIPTION
**Pull request by fspreiss**
Adds the source code of the threshold mock canister as example dapp.

### CODE REVIEW

File: rust/threshold-mock/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+crate-type = ["lib", "cdylib"]
+
+[dependencies]
+candid = "0.10.10"
+# Enable `custom` feature of indirect getrandom dependency. See lib.rs for more details.
+getrandom = { version = "0.2", features = ["custom"] }
+hex = "0.4"
+ic-cdk = "0.16.0"
+ic-cdk-timers = "0.10.0"
+ic-crypto-ed25519 = { git = "https://github.com/dfinity/ic" }
```
Comment (altkdf) [inline rust/threshold-mock/Cargo.toml:23]: should we use a particular `ic` revision?

File: rust/threshold-mock/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# Chain-key testing canister
+
+The chain-key testing canister is a canister smart contract that provides a [fake](https://www.martinfowler.com/articles/mocksArentStubs.html#TheDifferenceBetweenMocksAndStubs) implementation of the APIs of various chain-key-related features including threshold ECDSA, threshold Schnorr, and threshold key derivation (vetKeys) **for testing purposes**. The shortcut that this canister takes (compared to the production APIs) is that it relies on a cryptographic key that is hard-coded into the canister, rather than relying on a key that is distributed among the subnet nodes. With this, the implementation is inherently **insecure**: see [Disclaimer](#disclaimer).
```
Comment (altkdf) [inline rust/threshold-mock/README.md:3]: So in the link, there are two separate notions: a fake and a mock. We are using both here :)

File: rust/threshold-mock/README.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+* [Threshold Schnorr](https://internetcomputer.org/docs/current/developer-docs/smart-contracts/signatures/t-schnorr)
+    * [schnorr_public_key](https://internetcomputer.org/docs/current/references/ic-interface-spec#ic-schnorr_public_key), [sign_with_schnorr](https://internetcomputer.org/docs/current/references/ic-interface-spec#ic-sign_with_schnorr)
+* Preview: [Threshold Key Derivation (vetKeys)](https://internetcomputer.org/docs/current/developer-docs/smart-contracts/encryption/vetkeys)
+    * `vetkd_public_key`, `vetkd_encrypted_key` (see [API proposal PR](https://github.com/dfinity/interface-spec/pull/158))
+
+## Usage
+
+The canister is deployed on mainnet: the canister ID is `vrqyr-saaaa-aaaan-qzn4q-cai` ([dashboard](https://dashboard.internetcomputer.org/canister/vrqyr-saaaa-aaaan-qzn4q-cai), [candid-ui](https://a4gq6-oaaaa-aaaab-qaa4q-cai.raw.icp0.io/?id=vrqyr-saaaa-aaaan-qzn4q-cai)).
+
+All APIs support a single key ID: `insecure_test_key_1`.
```
Comment (altkdf) [inline rust/threshold-mock/README.md:19]: Might be useful to explain what's the difference with `test_key_1`

File: rust/threshold-mock/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    static MEMORY_MANAGER: RefCell<MemoryManager<DefaultMemoryImpl>> =
+        RefCell::new(MemoryManager::init(DefaultMemoryImpl::default()));
+    static CALL_COUNTS: RefCell<StableBTreeMap<String, u64, Memory>> = RefCell::new(
+            StableBTreeMap::init(
+                MEMORY_MANAGER.with_borrow(|m| m.get(MemoryId::new(1))),
+            )
+        );
+}
+
+const SEEDING_INTERVAL: Duration = Duration::from_secs(3600);
```
Comment (altkdf) [inline rust/threshold-mock/src/lib.rs:26]: what does seeding mean and why is it needed?

**Review by altkdf — APPROVED** [review]
Thanks @fspreiss , LGTM!

**Review by dfx-json — APPROVED** [review]

File: rust/threshold-mock/README.md (hunk 2 above)
Comment (fspreiss) [reply rust/threshold-mock/README.md:3]: Do you still see "mock" somewhere? I though I removed all places where "mock" was used in the README and across the repository.

File: rust/threshold-mock/src/lib.rs (hunk 4 above)
Comment (fspreiss) [reply rust/threshold-mock/src/lib.rs:26]: Seeding the underlying RNG, the code was taken 1:1 from https://internetcomputer.org/docs/current/developer-docs/security/security-best-practices/misc#2-using-raw_rand-as-seed-for-a-psuedo-random-number-generator-prng. I'll improve then naming to make clear the seeding is for the RNG.

Comment (fspreiss) [reply rust/threshold-mock/src/lib.rs:26]: https://github.com/dfinity/chainkey-testing-canister/pull/1

File: rust/threshold-mock/README.md (hunk 3 above)
Comment (fspreiss) [reply rust/threshold-mock/README.md:19]: https://github.com/dfinity/chainkey-testing-canister/pull/1

File: rust/threshold-mock/README.md (hunk 2 above)
Comment (altkdf) [reply rust/threshold-mock/README.md:3]: But the new repository name doesn't have `mock` in the name, so all good

**Pull request merged**